File: tools/new_game.py
#!/usr/bin/env python3
"""Bootstrap a new game bundle from an archive.

The platform's front door: turn *a link to a game archive* plus *one choice
(which executable starts the game)* into a ready-to-reconstruct bundle under
games/<name>/.

Why only one human input is needed -- the JIT-only config has just two kinds
of fields:

  * INPUT     program_path ... the one semantic choice (the entry exe)
  * DERIVED   name, runtime ... from the archive

There is no binaries list, entry symbol, or call-target table: the runtime takes
the entry cs:ip from the MZ header and JIT-compiles every reached segment on
demand. So this tool fetches + extracts the archive, detects the executables,
asks which one boots the game (only when it can't tell), writes the seed config,
and runs one probe build so you can see it link.

Usage:
    python -m tools.new_game <url|zip|dir> [--exe NAME] [--name NAME] [--no-probe]
"""
from __future__ import annotations

import argparse
import shutil
import sys
import tempfile
import urllib.parse
import urllib.request
import zipfile
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
GAMES_DIR = ROOT / "games"

try:
    from compiler.utils import sanitize_identifier
except ImportError:  # pragma: no cover - allow running as a script
    sys.path.insert(0, str(ROOT))
    from compiler.utils import sanitize_identifier

# Files that are never game runtime assets (docs / installer cruft).
_SKIP_EXT = {".txt", ".md", ".bat", ".diz", ".nfo"}

# Archive-packaging cruft that should never count as game content. These defeat
# the "single wrapping directory" heuristic (a __MACOSX/ sibling makes the
# top level look like it has two entries) and pollute the runtime file list.
_JUNK_NAMES = {"__macosx", ".ds_store", "thumbs.db", ".directory"}


def _is_junk(p: Path) -> bool:
    n = p.name.lower()
    return n in _JUNK_NAMES or n.startswith("._")


def _prune_junk(root: Path) -> None:
    """Delete archive-packaging cruft anywhere under *root*."""
    # Deepest-first so directories are emptied before removal.
    for p in sorted(root.rglob("*"), key=lambda x: len(x.parts), reverse=True):
        if not _is_junk(p):
            continue
        if p.is_dir():
            shutil.rmtree(p, ignore_errors=True)
        elif p.exists():
            p.unlink()


def fetch(src: str, workdir: Path) -> Path:
    """Resolve *src* (http(s) URL, local .zip, or local dir) to a local path."""
    if src.startswith(("http://", "https://")):
        name = Path(urllib.parse.urlparse(src).path).name or "archive.zip"
        dest = workdir / name
        print(f"fetch: downloading {src}")
        urllib.request.urlretrieve(src, dest)  # noqa: S310 - user-provided URL
        return dest
    p = Path(src).expanduser()
    if not p.exists():
        raise SystemExit(f"new-game: not found: {src}")
    return p


def _flatten_wrappers(dest: Path) -> None:
    """Drop a top-level wrapper directory when the archive holds nothing else.

    The common ``archive.zip -> archive/ -> files`` layout (and nested chains of
    such single-folder wrappers) leaves the game one or more directories deep,
    so the executable scan finds only an empty top level. Collapse it: while the
    top level's only real entry is a single directory, move that directory's
    contents up and remove it. Junk has already been pruned, so a ``__MACOSX/``
    sibling no longer makes the level look non-empty. Stops as soon as the top
    level holds a file or more than one entry -- we only ever drop a *pure*
    wrapper, never restructure real game content."""
    while True:
        entries = [e for e in dest.iterdir() if not _is_junk(e)]
        if len(entries) != 1 or not entries[0].is_dir():
            break
        sub = entries[0]
        # Rename first so a child sharing the wrapper's name can't collide.
        tmp = dest / (sub.name + ".__unwrap__")
        sub.rename(tmp)
        for item in list(tmp.iterdir()):
            shutil.move(str(item), str(dest / item.name))
        tmp.rmdir()


def extract_into(archive: Path, dest: Path) -> None:
    """Extract *archive* (zip), copy (dir), or place (single file) into *dest*.

    Prunes packaging cruft (__MACOSX/, .DS_Store, ...) then drops a pure
    top-level wrapper directory so the game's files sit directly in *dest*."""
    dest.mkdir(parents=True, exist_ok=True)
    if archive.is_dir():
        for item in archive.iterdir():
            target = dest / item.name
            if item.is_dir():
                shutil.copytree(item, target, dirs_exist_ok=True)
            else:
                shutil.copy2(item, target)
    elif zipfile.is_zipfile(archive):
        with zipfile.ZipFile(archive) as z:
            z.extractall(dest)
    else:
        shutil.copy2(archive, dest / archive.name)

    _prune_junk(dest)
    _flatten_wrappers(dest)


def is_executable(p: Path) -> bool:
    if p.suffix.lower() in (".exe", ".com"):
        return True
    try:
        with open(p, "rb") as f:
            return f.read(2) == b"MZ"
    except OSError:
        return False


class _cooked_terminal:
    """Force the controlling terminal into canonical ("cooked") mode for the
    duration of a prompt, then restore it.

    Earlier commands (the game, the FIFO control tools) can leave the tty in
    raw mode, where Enter sends a bare CR that input() never sees as a line
    end -- the user types a number, presses Enter, and just gets a literal
    ^M. Re-enabling ICANON/ECHO/ICRNL around the prompt makes it work no
    matter what state the terminal was left in. No-op when stdin is not a tty
    or termios is unavailable (e.g. Windows)."""

    def __enter__(self):
        self._saved = None
        try:
            import termios

            self._termios = termios
            fd = sys.stdin.fileno()
            self._fd = fd
            self._saved = termios.tcgetattr(fd)
            attrs = termios.tcgetattr(fd)
            attrs[0] |= termios.ICRNL          # map CR -> NL on input
            attrs[3] |= termios.ICANON | termios.ECHO | termios.ECHOE
            termios.tcsetattr(fd, termios.TCSANOW, attrs)
        except Exception:
            self._saved = None
        return self

    def __exit__(self, *exc):
        if self._saved is not None:
            try:
                self._termios.tcsetattr(
                    self._fd, self._termios.TCSANOW, self._saved
                )
            except Exception:
                pass
        return False


def choose_entry(execs: list[Path], requested: str | None) -> Path:
    """Pick the entry executable: --exe, the only one, or an interactive ask."""
    by_name = {e.name.lower(): e for e in execs}
    if requested:
        e = by_name.get(requested.lower())
        if not e:
            raise SystemExit(
                f"new-game: --exe {requested!r} not among executables: "
                + ", ".join(e.name for e in execs)
            )
        return e
    if len(execs) == 1:
        return execs[0]
    if not execs:
        raise SystemExit("new-game: no executables (.exe/.com/MZ) found in archive")
    if not sys.stdin.isatty():
        listing = ", ".join(e.name for e in execs)
        raise SystemExit(
            "new-game: multiple executables found; rerun with --exe NAME.\n"
            f"  candidates: {listing}"
        )
    print("\nWhich executable starts the game?")
    for i, e in enumerate(execs, 1):
        print(f"  {i}. {e.name}  ({e.stat().st_size} bytes)")
    # Accept either the list number or the executable name (case-insensitive),
    # and tolerate a stray CR/whitespace from a misbehaving terminal.
    with _cooked_terminal():
        while True:
            try:
                raw = input("Enter number (or exe name): ").strip().strip("\r")
            except EOFError:
                raise SystemExit(
                    "new-game: no input; rerun with --exe NAME to pick "
                    "non-interactively."
                )
            if raw.isdigit() and 1 <= int(raw) <= len(execs):
                return execs[int(raw) - 1]
            if raw.lower() in by_name:
                return by_name[raw.lower()]
            print(f"  (invalid: type 1-{len(execs)} or one of: "
                  + ", ".join(e.name for e in execs) + ")")


def build_seed_config(name: str, exe: Path, game_dir: Path) -> dict:
    runtime = []
    # Recurse: a DOS game opens its assets by path (e.g. fopen("DATA\\GFX.DAT")),
    # so subdirectories are part of the data layout, not packaging. Preserve each
    # asset's path relative to the bundle as its dest -- the runtime resolves the
    # backslash path to that same subdir (resolve_case_insensitive_path), and the
    # copy step recreates the tree under build/<game>/.
    for f in sorted(game_dir.rglob("*"), key=lambda p: str(p).lower()):
        if not f.is_file() or _is_junk(f):
            continue
        if f.suffix.lower() in _SKIP_EXT:
            continue
        rel = f.relative_to(game_dir).as_posix()
        if rel == f"{name}.json":
            continue
        runtime.append({"source": f"games/{name}/{rel}", "dest": rel})
    # JIT-only: the config just names the program image and the data files to
    # copy. The runtime takes the entry cs:ip from the MZ header and JIT-compiles
    # every reached segment on demand -- no binaries list, no entry symbol, no
    # call-target table to grow. A game needing a non-default PSP load segment
    # sets "psp_seg" here later.
    return {
        "name": name,
        "program_path": exe.name,
        "runtime": runtime,
    }


def probe(name: str) -> int:
    """Run one game-agnostic build pass and report how far it got."""
    import subprocess

    print(f"\nprobe: building bundle '{name}' (emit config + link runtime)...")
    proc = subprocess.run(
        [sys.executable, "-m", "tools.game", "build", name],
        cwd=ROOT,
    )
    if proc.returncode == 0:
        print(
            f"\nprobe: OK -- bundle '{name}' builds.\n"
            f"  next: python -m tools.game run {name} --headless   to see it boot.\n"
            f"  The JIT compiles the game's code on demand as it runs; watch the\n"
            f"  screenshots + trace to see how far it gets."
        )
    else:
        print(
            f"\nprobe: build stopped (exit {proc.returncode}) -- inspect the output\n"
            f"  above. A build failure here is a config/link error (the build only\n"
            f"  emits the config and links the runtime; no game code is decoded)."
        )
    return proc.returncode


def main() -> None:
    ap = argparse.ArgumentParser(
        prog="new-game", description="Bootstrap a game bundle from an archive."
    )
    ap.add_argument("archive", help="http(s) URL, local .zip, or local directory")
    ap.add_argument("--exe", help="entry executable name (skips the prompt)")
    ap.add_argument("--name", help="bundle name (default: archive stem)")
    ap.add_argument(
        "--no-probe", action="store_true", help="scaffold only; skip the build probe"
    )
    ap.add_argument(
        "--force", action="store_true", help="overwrite games/<name>/ if it exists"
    )
    args = ap.parse_args()

    name = sanitize_identifier(args.name or Path(args.archive.rstrip("/")).stem)
    game_dir = GAMES_DIR / name
    if game_dir.exists():
        if not args.force:
            raise SystemExit(
                f"new-game: games/{name}/ already exists (use --force to overwrite)"
            )
        shutil.rmtree(game_dir)

    with tempfile.TemporaryDirectory() as tmp:
        archive = fetch(args.archive, Path(tmp))
        print(f"extract: -> games/{name}/")
        extract_into(archive, game_dir)

    execs = sorted(
        (f for f in game_dir.iterdir() if f.is_file() and is_executable(f)),
        key=lambda p: p.name.lower(),
    )
    print(f"detect: {len(execs)} executable(s): " + ", ".join(e.name for e in execs))
    exe = choose_entry(execs, args.exe)

    config = build_seed_config(name, exe, game_dir)
    config_path = game_dir / f"{name}.json"
    import json

    config_path.write_text(json.dumps(config, indent=2) + "\n")
    print(
        f"scaffold: wrote {config_path.relative_to(ROOT)}\n"
        f"  program_path  = {config['program_path']}\n"
        f"  entry_symbol  = {config['entry_symbol']}\n"
        f"  binaries      = {len(config['binaries'])} (entry exe; loop adds more)\n"
        f"  runtime files = {len(config['runtime'])}"
    )

    if args.no_probe:
        return
    sys.exit(probe(name))


if __name__ == "__main__":
    main()

File: saisei/src/new_game.rs
//! `saisei new-game` — bootstrap a game bundle from an archive (URL / .zip / dir).
//! Downloading and unzipping happen *in-process* (ureq + the zip crate), so a
//! fresh clone needs no curl and no unzip on PATH — `cargo build` is the whole
//! install. Fetches + extracts, detects executables, picks the entry exe, writes
//! the seed config, and runs one probe build.

use serde_json::{json, Value};
use std::io::Write;
use std::path::{Path, PathBuf};
use std::process::exit;

fn die(msg: &str) -> ! {
    eprintln!("new-game: {msg}");
    exit(1)
}

const SKIP_EXT: &[&str] = &["txt", "md", "bat", "diz", "nfo"];
const JUNK_NAMES: &[&str] = &["__macosx", ".ds_store", "thumbs.db", ".directory"];

fn is_junk(p: &Path) -> bool {
    let n = p
        .file_name()
        .map(|s| s.to_string_lossy().to_lowercase())
        .unwrap_or_default();
    JUNK_NAMES.contains(&n.as_str()) || n.starts_with("._")
}

fn walk(dir: &Path, out: &mut Vec<PathBuf>) {
    if let Ok(rd) = std::fs::read_dir(dir) {
        for e in rd.flatten() {
            let p = e.path();
            out.push(p.clone());
            if p.is_dir() {
                walk(&p, out);
            }
        }
    }
}

fn prune_junk(root: &Path) {
    let mut all = Vec::new();
    walk(root, &mut all);
    all.sort_by_key(|p| std::cmp::Reverse(p.components().count()));
    for p in all {
        if !is_junk(&p) {
            continue;
        }
        if p.is_dir() {
            std::fs::remove_dir_all(&p).ok();
        } else if p.exists() {
            std::fs::remove_file(&p).ok();
        }
    }
}

fn flatten_wrappers(dest: &Path) {
    loop {
        let entries: Vec<PathBuf> = match std::fs::read_dir(dest) {
            Ok(rd) => rd
                .flatten()
                .map(|e| e.path())
                .filter(|p| !is_junk(p))
                .collect(),
            Err(_) => return,
        };
        if entries.len() != 1 || !entries[0].is_dir() {
            break;
        }
        let sub = &entries[0];
        let tmp = dest.join(format!(
            "{}.__unwrap__",
            sub.file_name().unwrap().to_string_lossy()
        ));
        std::fs::rename(sub, &tmp).ok();
        if let Ok(rd) = std::fs::read_dir(&tmp) {
            for item in rd.flatten() {
                std::fs::rename(item.path(), dest.join(item.file_name())).ok();
            }
        }
        std::fs::remove_dir_all(&tmp).ok();
    }
}

fn copy_dir_into(src: &Path, dest: &Path) {
    for item in std::fs::read_dir(src).into_iter().flatten().flatten() {
        let target = dest.join(item.file_name());
        if item.path().is_dir() {
            std::fs::create_dir_all(&target).ok();
            copy_dir_into(&item.path(), &target);
        } else {
            std::fs::copy(item.path(), &target).ok();
        }
    }
}

fn is_zip(p: &Path) -> bool {
    std::fs::File::open(p)
        .and_then(|mut f| {
            let mut b = [0u8; 4];
            use std::io::Read;
            f.read_exact(&mut b)?;
            Ok(b == *b"PK\x03\x04")
        })
        .unwrap_or(false)
}

/// Download `src` into `workdir`. HTTP(S) is served in-process by ureq —
/// redirects followed, TLS via rustls, so there is no curl, no OpenSSL, and no
/// system CA bundle in the picture.
fn download(src: &str, workdir: &Path) -> Result<PathBuf, String> {
    // Name the download after the URL's last path segment, minus any ?query#frag.
    let name = src
        .rsplit('/')
        .next()
        .map(|s| s.split(['?', '#']).next().unwrap_or(s))
        .filter(|s| !s.is_empty())
        .unwrap_or("archive.zip");
    let dest = workdir.join(name);

    let resp = ureq::get(src)
        .call()
        .map_err(|e| format!("Could not download {src}: {e}"))?;
    // No read limit: game archives run to tens of MB and ureq's default body cap
    // would silently truncate them.
    let mut body = resp.into_body().into_with_config().limit(u64::MAX).reader();
    let mut out =
        std::fs::File::create(&dest).map_err(|e| format!("create {}: {e}", dest.display()))?;
    std::io::copy(&mut body, &mut out).map_err(|e| format!("Could not download {src}: {e}"))?;
    Ok(dest)
}

/// Unpack `archive` into `dest`. A zip is extracted in-process (no unzip on
/// PATH); a directory or a bare game executable is copied. Anything else is
/// refused by name — better a one-line reason than a bundle that silently holds
/// an unextracted .7z.
fn extract_into(archive: &Path, dest: &Path) {
    std::fs::create_dir_all(dest).ok();
    if archive.is_dir() {
        copy_dir_into(archive, dest);
    } else if is_zip(archive) {
        extract_zip(archive, dest);
    } else if is_executable(archive) {
        std::fs::copy(archive, dest.join(archive.file_name().unwrap()))
            .unwrap_or_else(|e| die(&format!("copy {}: {e}", archive.display())));
    } else {
        die(&format!(
            "{}: not a zip.\n  \
             Only .zip archives are supported for now. Unpack it yourself and \
             point new-game at the folder:\n  saisei new-game <folder>",
            archive.display()
        ));
    }
    prune_junk(dest);
    flatten_wrappers(dest);
}

/// Extract a zip with the `zip` crate: stored + deflate, which is what every
/// archive site serves. A 1990s PKZIP-imploded file would land here as an
/// unsupported-method error naming the entry, not as a corrupt bundle.
fn extract_zip(archive: &Path, dest: &Path) {
    let f = std::fs::File::open(archive)
        .unwrap_or_else(|e| die(&format!("open {}: {e}", archive.display())));
    let mut zip = zip::ZipArchive::new(f)
        .unwrap_or_else(|e| die(&format!("{}: not a readable zip: {e}", archive.display())));

    for i in 0..zip.len() {
        let mut entry = zip.by_index(i).unwrap_or_else(|e| {
            die(&format!(
                "{}: unreadable zip entry #{i}: {e}\n  \
                 (only stored and deflate compression are supported)",
                archive.display()
            ))
        });
        // enclosed_name() rejects absolute paths and `..` traversal, so a hostile
        // or just-malformed zip can never write outside games/<name>/.
        let Some(rel) = entry.enclosed_name() else {
            eprintln!("new-game: skipping unsafe path in zip: {}", entry.name());
            continue;
        };
        let out = dest.join(rel);
        if entry.is_dir() {
            std::fs::create_dir_all(&out).ok();
            continue;
        }
        if let Some(parent) = out.parent() {
            std::fs::create_dir_all(parent).ok();
        }
        let mut w = std::fs::File::create(&out)
            .unwrap_or_else(|e| die(&format!("create {}: {e}", out.display())));
        std::io::copy(&mut entry, &mut w).unwrap_or_else(|e| {
            die(&format!(
                "{}: extracting {}: {e}\n  \
                 (only stored and deflate compression are supported)",
                archive.display(),
                entry.name()
            ))
        });
    }
}

pub fn is_executable(p: &Path) -> bool {
    let ext = p
        .extension()
        .map(|s| s.to_string_lossy().to_lowercase())
        .unwrap_or_default();
    if ext == "exe" || ext == "com" {
        return true;
    }
    std::fs::File::open(p)
        .and_then(|mut f| {
            let mut b = [0u8; 2];
            use std::io::Read;
            f.read_exact(&mut b)?;
            Ok(&b == b"MZ")
        })
        .unwrap_or(false)
}

fn choose_entry(execs: &[PathBuf], requested: Option<&str>) -> PathBuf {
    let nm = |p: &Path| p.file_name().unwrap().to_string_lossy().to_string();
    if let Some(req) = requested {
        for e in execs {
            if nm(e).to_lowercase() == req.to_lowercase() {
                return e.clone();
            }
        }
        die(&format!(
            "--exe {req:?} not among executables: {}",
            execs.iter().map(|e| nm(e)).collect::<Vec<_>>().join(", ")
        ));
    }
    if execs.len() == 1 {
        return execs[0].clone();
    }
    if execs.is_empty() {
        die("no executables (.exe/.com/MZ) found in archive");
    }
    use std::io::IsTerminal;
    if !std::io::stdin().is_terminal() {
        die(&format!(
            "multiple executables found; rerun with --exe NAME.\n  candidates: {}",
            execs.iter().map(|e| nm(e)).collect::<Vec<_>>().join(", ")
        ));
    }
    println!("\nWhich executable starts the game?");
    for (i, e) in execs.iter().enumerate() {
        let sz = e.metadata().map(|m| m.len()).unwrap_or(0);
        println!("  {}. {}  ({sz} bytes)", i + 1, nm(e));
    }
    loop {
        print!("Enter number (or exe name): ");
        std::io::stdout().flush().ok();
        let mut line = String::new();
        if std::io::stdin().read_line(&mut line).unwrap_or(0) == 0 {
            die("no input; rerun with --exe NAME to pick non-interactively.");
        }
        let raw = line.trim();
        if let Ok(n) = raw.parse::<usize>() {
            if n >= 1 && n <= execs.len() {
                return execs[n - 1].clone();
            }
        }
        for e in execs {
            if nm(e).to_lowercase() == raw.to_lowercase() {
                return e.clone();
            }
        }
        println!(
            "  (invalid: type 1-{} or one of: {})",
            execs.len(),
            execs.iter().map(|e| nm(e)).collect::<Vec<_>>().join(", ")
        );
    }
}

pub fn build_seed_config(name: &str, exe: &Path, game_dir: &Path) -> Value {
    let mut files = Vec::new();
    walk(game_dir, &mut files);
    files.sort_by_key(|p| p.to_string_lossy().to_lowercase());
    let mut runtime = Vec::new();
    for f in &files {
        if !f.is_file() || is_junk(f) {
            continue;
        }
        let ext = f
            .extension()
            .map(|s| s.to_string_lossy().to_lowercase())
            .unwrap_or_default();
        if SKIP_EXT.contains(&ext.as_str()) {
            continue;
        }
        let rel = f
            .strip_prefix(game_dir)
            .unwrap()
            .to_string_lossy()
            .replace('\\', "/");
        if rel == format!("{name}.json") {
            continue;
        }
        runtime.push(json!({ "source": format!("games/{name}/{rel}"), "dest": rel }));
    }
    json!({
        "name": name,
        "program_path": exe.file_name().unwrap().to_string_lossy(),
        "runtime": runtime,
    })
}

/// A bundle unpacked into `games/<name>/`, with the executables found in it.
///
/// The only question left is which of them starts the game — which the CLI asks
/// on a tty and the player app asks with a list of tiles.
#[derive(Debug)]
pub struct Staged {
    pub name: String,
    pub game_dir: PathBuf,
    pub execs: Vec<PathBuf>,
}

impl Staged {
    /// The executables by bare file name, for showing to a person.
    pub fn exe_names(&self) -> Vec<String> {
        self.execs
            .iter()
            .map(|e| e.file_name().unwrap().to_string_lossy().into_owned())
            .collect()
    }
}

/// Fetch `src` (a URL, a zip, or a folder) and unpack it into `games/<name>/`.
///
/// Returns errors rather than exiting: the CLI can print them and stop, but the
/// player app is a window and has to be able to say "that didn't work" and carry
/// on.
pub fn stage(
    root: &Path,
    src: &str,
    name_hint: Option<&str>,
    force: bool,
) -> Result<Staged, String> {
    let stem = Path::new(src.trim_end_matches('/'))
        .file_stem()
        .map(|s| s.to_string_lossy().into_owned())
        .unwrap_or_default();
    let name = crate::sanitize_identifier(&name_hint.map(str::to_string).unwrap_or(stem));
    if name.is_empty() {
        return Err("could not work out a name for this game".into());
    }
    let game_dir = root.join("games").join(&name);
    if game_dir.exists() {
        if !force {
            return Err(format!("You already have a game called '{name}'."));
        }
        std::fs::remove_dir_all(&game_dir).ok();
    }

    // Refuse what we cannot unpack *before* unpacking, so the extract path has
    // nothing left to fail on.
    let is_url = src.starts_with("http://") || src.starts_with("https://");
    if !is_url {
        let p = Path::new(src);
        if !p.exists() {
            return Err(format!("not found: {src}"));
        }
        if !p.is_dir() && !is_zip(p) && !is_executable(p) {
            return Err("Only .zip archives and folders can be added.".into());
        }
    }

    let tmp = std::env::temp_dir().join(format!("saisei_newgame_{name}"));
    std::fs::create_dir_all(&tmp).ok();
    let fetched = if is_url {
        download(src, &tmp)?
    } else {
        PathBuf::from(src)
    };
    extract_into(&fetched, &game_dir);
    std::fs::remove_dir_all(&tmp).ok();

    let mut execs = Vec::new();
    walk(&game_dir, &mut execs);
    execs.retain(|p| p.is_file() && is_executable(p));
    execs.sort_by_key(|p| p.file_name().unwrap().to_string_lossy().to_lowercase());
    if execs.is_empty() {
        std::fs::remove_dir_all(&game_dir).ok();
        return Err("No DOS program (.exe/.com) in there.".into());
    }

    Ok(Staged {
        name,
        game_dir,
        execs,
    })
}

/// Write the bundle's config, naming `exe` as the program to run.
pub fn finish(staged: &Staged, exe: &Path) -> Result<PathBuf, String> {
    let config = build_seed_config(&staged.name, exe, &staged.game_dir);
    let config_path = staged.game_dir.join(format!("{}.json", staged.name));
    std::fs::write(
        &config_path,
        format!("{}\n", serde_json::to_string_pretty(&config).unwrap()),
    )
    .map_err(|e| format!("write {}: {e}", config_path.display()))?;
    Ok(config_path)
}

pub fn main(root: &Path, args: &[String]) -> ! {
    let mut archive: Option<String> = None;
    let mut exe: Option<String> = None;
    let mut name_arg: Option<String> = None;
    let mut no_probe = false;
    let mut force = false;
    let mut it = args.iter();
    while let Some(a) = it.next() {
        match a.as_str() {
            "--exe" => exe = it.next().cloned(),
            "--name" => name_arg = it.next().cloned(),
            "--no-probe" => no_probe = true,
            "--force" => force = true,
            other if other.starts_with("--") => die(&format!("unknown flag: {other}")),
            other => archive = Some(other.to_string()),
        }
    }
    let archive = archive.unwrap_or_else(|| die("usage: saisei new-game <url|zip|dir> [--exe NAME] [--name NAME] [--force] [--no-probe]"));

    let staged = stage(root, &archive, name_arg.as_deref(), force).unwrap_or_else(|e| die(&e));
    let name = staged.name.clone();
    println!("extract: -> games/{name}/");
    let exe_path = choose_entry(&staged.execs, exe.as_deref());
    println!(
        "detect: {} executable(s): {}",
        staged.execs.len(),
        staged.exe_names().join(", ")
    );

    finish(&staged, &exe_path).unwrap_or_else(|e| die(&e));
    println!(
        "scaffold: wrote games/{name}/{name}.json\n  program_path  = {}",
        exe_path.file_name().unwrap().to_string_lossy()
    );

    if no_probe {
        exit(0);
    }
    println!("\nprobe: building bundle '{name}' (emit config + link runtime)...");
    let game = crate::load_game_definition(root, &name, None);
    crate::build(root, &game); // dies on build failure
    println!("\nprobe: OK -- bundle '{name}' builds.\n  next: saisei-cli run {name} --headless   to see it boot.");
    exit(0);
}
